ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/ConversationIndexMemory.java

```diff
@@ -0,0 +1,215 @@
+import org.opensearch.ml.memory.action.conversation.CreateInteractionResponse;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.search.sort.SortOrder;
+
+import lombok.Getter;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@Getter
+public class ConversationIndexMemory implements Memory {
+    public static final String TYPE = "conversation_index";
+    public static final String CONVERSATION_ID = "conversation_id";
+    public static final String FINAL_ANSWER = "final_answer";
+    public static final String CREATED_TIME = "created_time";
+    public static final String MEMORY_NAME = "memory_name";
+    public static final String MEMORY_ID = "memory_id";
+    public static final String APP_TYPE = "app_type";
+    public static int LAST_N_INTERACTIONS = 10;
+    protected String memoryMetaIndexName;
+    protected String memoryMessageIndexName;
+    protected String conversationId;
+    protected boolean retrieveFinalAnswer = true;
+    protected final Client client;
+    private final MLIndicesHandler mlIndicesHandler;
+    private MLMemoryManager memoryManager;
+
+    public ConversationIndexMemory(
+        Client client,
+        MLIndicesHandler mlIndicesHandler,
+        String memoryMetaIndexName,
+        String memoryMessageIndexName,
+        String conversationId,
+        MLMemoryManager memoryManager
+    ) {
+        this.client = client;
+        this.mlIndicesHandler = mlIndicesHandler;
+        this.memoryMetaIndexName = memoryMetaIndexName;
+        this.memoryMessageIndexName = memoryMessageIndexName;
+        this.conversationId = conversationId;
+        this.memoryManager = memoryManager;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public void save(String id, Message message) {
+        this.save(id, message, ActionListener.wrap(r -> { log.info("saved message into {} memory, session id: {}", TYPE, id); }, e -> {
+            log.error("Failed to save message to memory", e);
+        }));
+    }
+
+    @Override
+    public void save(String id, Message message, ActionListener listener) {
+        mlIndicesHandler.initMemoryMessageIndex(ActionListener.wrap(created -> {
+            if (created) {
+                IndexRequest indexRequest = new IndexRequest(memoryMessageIndexName);
+                ConversationIndexMessage conversationIndexMessage = (ConversationIndexMessage) message;
+                XContentBuilder builder = XContentBuilder.builder(XContentType.JSON.xContent());
+                conversationIndexMessage.toXContent(builder, ToXContent.EMPTY_PARAMS);
+                indexRequest.source(builder);
+                client.index(indexRequest, listener);
+            } else {
+                listener.onFailure(new RuntimeException("Failed to create memory message index"));
+            }
+        }, e -> { listener.onFailure(new RuntimeException("Failed to create memory message index")); }));
```

Comment:
better to keep the original cause
```suggestion
        }, e -> { listener.onFailure(new RuntimeException("Failed to create memory message index", e)); }));
```

Reply:
makes sense, will change